File: Cpp_Collections_benchmark/Benchmark.cpp
#include "Benchmark.h"
#include <chrono>
#include <iostream>
#include <iomanip>


Benchmark::Benchmark(IBenchmark* benchmark)
{
	this->benchmark = benchmark;
}

BenchmarkResults* Benchmark::PerformAllTests()
{
	PerformAdd();
	PerformFind();
	PerformRemove();

	return &this->results;
}

void Benchmark::PerformAdd()
{
	auto start = std::chrono::high_resolution_clock::now();
	benchmark->addTest();
	auto end = std::chrono::high_resolution_clock::now();

	auto elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() / 1000000.0;
	std::cout << elapsed << std::endl;
	results.AddTime = elapsed;
}

void Benchmark::PerformFind()
{
	auto start = std::chrono::high_resolution_clock::now();
	benchmark->findTest();
	auto end = std::chrono::high_resolution_clock::now();
	
	auto elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() / 1000000.0;
	std::cout << elapsed << std::endl;
	results.FindTime = elapsed;
}

void Benchmark::PerformRemove()
{
	auto start = std::chrono::high_resolution_clock::now();
	benchmark->removeTest();
	auto end = std::chrono::high_resolution_clock::now();
	auto elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count()/1000000.0;
	std::cout << elapsed << std::endl;
	results.RemoveTime = elapsed;
}

File: Java_Collections_Benchmark/src/com/company/Benchmark.java
package com.company;


public class Benchmark {

    IBenchmark benchmark;
    BenchmarkResults results;

    public Benchmark(IBenchmark benchmark){
        this.benchmark = benchmark;
        results = new BenchmarkResults();
    }

    public BenchmarkResults PerformAllTests(){

        PerformAdd();
        PerformFind();
        PerformRemove();
        return results;
    }

    private void PerformAdd(){

        long start = System.nanoTime();
        benchmark.AddTest();
        long end = System.nanoTime();

        results.AddTime = (end - start)/1000000.0;
    }

    private void PerformFind(){

        long start = System.nanoTime();
        benchmark.FindTest();
        long end = System.nanoTime();

        results.FindTime = (end - start)/1000000.0;
    }

    private void PerformRemove(){

        long start = System.nanoTime();
        benchmark.RemoveTest();
        long end = System.nanoTime();

        results.RemoveTime = (end - start)/1000000.0;
    }

}
